lib/airbrake/rails/app.rb

```diff
@@ -39,13 +39,7 @@ def self.recognize_route(request)
 
             # Skip "catch-all" routes such as:
             #   get '*path => 'pages#about'
-            #
-            # @todo The `glob?` method was added in Rails v4.2.0.beta1. We
-            # should remove the `respond_to?` check once we drop old Rails
-            # versions support.
-            #
-            # https://github.com/rails/rails/commit/5460591f0226a9d248b7b4f89186bd5553e7768f
-            next if route.respond_to?(:glob?) && route.glob?
+            next if route.path.ast.nil? && route.glob?
```

Comment:
With this version I'm getting the following error in my rails 7.0.1 apps (`with CI=true` eager loading the code):

```
  9) GET /api/v1/user without a valid api token returns with status code 401
     Failure/Error: get '/api/v1/user'

     NoMethodError:
       undefined method `glob?' for nil:NilClass

               path.ast.glob?
                       ^^^^^^
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/bundler/gems/airbrake-601193db2c93/lib/airbrake/rails/app.rb:42:in `block (2 levels) in recognize_route'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/bundler/gems/airbrake-601193db2c93/lib/airbrake/rails/app.rb:35:in `block in recognize_route'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/bundler/gems/airbrake-601193db2c93/lib/airbrake/rails/app.rb:34:in `each'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/bundler/gems/airbrake-601193db2c93/lib/airbrake/rails/app.rb:34:in `recognize_route'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/bundler/gems/airbrake-601193db2c93/lib/airbrake/rails/action_controller_route_subscriber.rb:20:in `call'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/warden-1.2.9/lib/warden/manager.rb:36:in `block in call'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/warden-1.2.9/lib/warden/manager.rb:34:in `catch'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/warden-1.2.9/lib/warden/manager.rb:34:in `call'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/rack-2.2.3/lib/rack/tempfile_reaper.rb:15:in `call'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/rack-2.2.3/lib/rack/etag.rb:27:in `call'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/rack-2.2.3/lib/rack/conditional_get.rb:27:in `call'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/rack-2.2.3/lib/rack/head.rb:12:in `call'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:266:in `context'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:260:in `call'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/bundler/gems/airbrake-601193db2c93/lib/airbrake/rack/middleware.rb:34:in `call!'
     # /Users/meyer/.rvm/gems/ruby-3.1.0@whatchy/bundler/gems/airbrake-601193db2c93/lib/airbrake/rack/middleware.rb:23:in `call'
```

The error is actually coming from this [LOC](https://github.com/rails/rails/blob/9fb6cdf8fafea8d23fae06586f0f3248f873d53c/actionpack/lib/action_dispatch/journey/route.rb#L138) in rails:

```ruby
def glob?
  path.ast.glob?
end
```

Shouldn't the check therefore be like this instead?

```diff
- next if route.path.ast.nil? && route.glob?
+ next if !route.path.ast.nil? && route.glob?
```

Perhaps it would be a good idea to add a "catch-all" to the [routes in the dummy rails app](https://github.com/airbrake/airbrake/blob/4d7cacb7c3cef6b3d7f4cecb31bc16afeee7b832/spec/apps/rails/dummy_app.rb#L40-L60)?

Let me know if I can help somehow.

Reply:
I forgot to submit my review. It should be

```suggestion
            next if route.path.ast.nil? || route.glob?
```

Either way, I had a chance to speak to someone who was modifying the code in Rails and we need a different approach here. I will submit a PR and release a new version. @klausmeyer would you be able to test my change against your app prior to releasing a new version?

Reply:
Yes sure. Just ping me and I'll give it a try.

Reply:
Thank you very much @klausmeyer. Here's the PR: https://github.com/airbrake/airbrake/pull/1211
Branch: `1177-rails-7-glob-fix`

Reply:
👀 I'm on it

Reply:
✅ RSpec Suite of my 7.0.1 app with `config.eager_load = true`
✅ Deployment to Heroku
✅ Error reporting with `rake airbrake:test`

Used versions:

```
$ grep -A2 airbrake Gemfile.lock
  remote: https://github.com/airbrake/airbrake.git
  revision: 7e03129eef591c46cce1c465533c3efcab03d09e
  branch: 1177-rails-7-glob-fix
--
    airbrake (12.0.0)
      airbrake-ruby (~> 6.0)

GIT
--
    airbrake-ruby (6.0.2)
      rbtree3 (~> 0.5)
    autoprefixer-rails (10.3.3.0)
--
  airbrake!
  bootsnap
  bootstrap (~> 4.3)
```

Reply:
Thanks a bunch!